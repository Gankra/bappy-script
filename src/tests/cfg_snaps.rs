fn dump_cfg(input: &str) -> String {
    let mut program = crate::Program::typed(input);
    program.parse().unwrap();
    program.check();

    let ctx = program.ctx.unwrap();
    let cfg = program.cfg.unwrap();

    cfg.format(&ctx).unwrap()
}

#[test]
fn capture_basics() {
    let program = r#"
    struct Point {
        x: Int,
        y: Int,
    }
    
    let pt = Point { x: 2, y: 7 }
    fn captures(arg: Int) -> Int {
        ret add(arg, pt.y)
    }

    fn no_captures(arg: Int) -> Int {
        ret add(arg, arg)
    }

    fn calls_funcs() -> Int {
        let a = captures(23)
        let b = no_captures(29)
        ret add(a, b)
    }

    let my_fn = captures
    let my_fn2 = no_captures

    print captures(2)
    print my_fn(3)

    print no_captures(6)
    print my_fn2(7)

    print calls_funcs()

    let x = 1
    let y = true
    let z = (x, y)
    if z.1 {
        print add(1, add(x, 2))
    }
    ret 0
"#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("basics", cfg);
}

#[test]
fn loop_basics() {
    let program = r#"
    let cond = true
    let mut counter = 3
    loop {
        if eq(counter, 0) {
            break
        }
        set counter = sub(counter, 1)
        print counter

        loop {
            print "hello"
            if cond {
                break
            } else {
                continue
            }
            print "unreachable"
        }
        print "looped!"
    }
    ret 0
"#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("loop_basics", cfg);
}

#[test]
fn func_names_are_paths() {
    let program = r#"
    fn do_a_compy(x: Int) -> Int {
        ret mul(x, x)
    }
    let z = 5
    fn captured_do_a_compy(x: Int) -> Int {
        ret do_a_compy(add(x, z))
    }
    struct MyClosure {
        func: fn(Int) -> Int
        capture: Int
    }
    fn call_closure(close: MyClosure) -> Int {
        ret close.func(close.capture)
    }
    
    let close1 = MyClosure { func: do_a_compy, capture: 7 }
    let close2 = MyClosure { func: captured_do_a_compy, capture: 9 }
    
    print call_closure(close1)
    print call_closure(close2)
    
    ret close2.capture
    "#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("func_names_are_paths", cfg);
}
