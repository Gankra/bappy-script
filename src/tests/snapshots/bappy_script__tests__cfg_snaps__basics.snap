---
source: src/tests/cfg_snaps.rs
assertion_line: 56
expression: cfg

---
#struct0_Point {
  x: Int
  y: Int
}

[intrinsic] #fn0_add(lhs, rhs)
[intrinsic] #fn1_sub(lhs, rhs)
[intrinsic] #fn2_mul(lhs, rhs)
[intrinsic] #fn3_eq(lhs, rhs)
[intrinsic] #fn4_not(rhs)

#fn5_main():
  bb0():
    %0 = Int(2)
    %1 = Int(7)
    %2 = #struct0_Point{ x: %0, y: %1 }
    %3 = (%2,)
    %4 = heap_alloc()
    %4.* = %3
    %5 = [closure](#fn6_captures, %4)
    %6 = (%5,)
    %7 = heap_alloc()
    %7.* = %6
    %8 = [closure](#fn8_calls_funcs, %7)
    %9 = #fn7_no_captures
    %10 = Int(2)
    %11 = (%5)(%10)
    print %11
    %12 = Int(3)
    %13 = (%5)(%12)
    print %13
    %14 = Int(6)
    %15 = #fn7_no_captures(%14)
    print %15
    %16 = Int(7)
    %17 = (%9)(%16)
    print %17
    %18 = (%8)()
    print %18
    %19 = Int(1)
    %20 = Bool(true)
    %21 = (%19, %20)
    %22 = %21.1
    cond %22: bb1(), bb2()

  bb1():
    %23 = Int(1)
    %24 = Int(2)
    %25 = #fn0_add(%19, %24)
    %26 = #fn0_add(%23, %25)
    print %26
    jmp bb3()

  bb2():
    jmp bb3()

  bb3():
    %27 = Int(0)
    ret %27


#fn6_captures(%0_arg: Int, %4_[closure]: &(Point,)):
  bb0(%0, %4):
    %1 = %4.*.0
    %2 = %1.y
    %3 = #fn0_add(%0, %2)
    ret %3


#fn7_no_captures(%0_arg: Int):
  bb0(%0):
    %1 = #fn0_add(%0, %0)
    ret %1


#fn8_calls_funcs(%6_[closure]: &(fn (Int) -> Int,)):
  bb0(%6):
    %0 = %6.*.0
    %1 = Int(23)
    %2 = (%0)(%1)
    %3 = Int(29)
    %4 = #fn7_no_captures(%3)
    %5 = #fn0_add(%2, %4)
    ret %5


